File: src/test_priorityq.py
"""Functions that test Priority class methods."""
import pytest


@pytest.fixture
def new_pq():
    """Create an instance of a priority queue."""
    from priorityq import Priority
    pq = Priority()
    return pq


def test_new_priority_instance_empty_upon_init(new_pq):
    """Return value added by insert method."""
    assert new_pq.head is None
    assert new_pq.tail is None
    assert new_pq._counter == 0
    assert new_pq.priorities == []


def test_insert_method_adds_value_to_priority_queue(new_pq):
    """Return value added by insert method."""
    new_pq.insert(1)
    assert new_pq.head.data == 1


def test_insert_method_adds_value_to_end_of_pq(new_pq):
    """Return value added by insert method."""
    new_pq.insert(1)
    new_pq.insert(2)
    assert new_pq.tail.data == 2
    assert new_pq.head.data == 1
    assert new_pq.head.next_node.data == 2


def test_insert_method_adds_value_and_priority_to_node_in_pq(new_pq):
    """Return value added by insert method."""
    new_pq.insert(1, 100)
    assert new_pq.tail.data == 1
    assert new_pq.head.data == 1
    assert new_pq.head.priority == 100


def test_pop_method_removes_and_returns_value_from_head_of_pq(new_pq):
    """Delete and return value from head of priority queue."""
    new_pq.insert(1)
    assert new_pq.pop() == 1


def test_pop_method_removes_and_returns_only_head_of_pq(new_pq):
    """Delete and return value from head of priority queue."""
    new_pq.insert(1)
    new_pq.insert(2)
    new_pq.insert(3)
    assert new_pq.pop() == 1


def test_pop_method_removes_and_returns_highest_priority_of_pq(new_pq):
    """Delete and return value from head of priority queue."""
    new_pq.insert(1, 100)
    new_pq.insert(2, 99.9)
    new_pq.insert(3, 99.8)
    assert new_pq.pop() == 1


def test_pop_method_raises_exception_when_queue_is_empty(new_pq):
    """Delete and return value from head of priority queue."""
    with pytest.raises(IndexError):
        assert new_pq.pop() == 1


def test_peek_method_returns_none_if_pq_is_empty(new_pq):
    """Raise IndexError if priority queue is empty."""
    with pytest.raises(IndexError):
        assert new_pq.peek()


def test_peek_method_returns_value_of_highest_priority_item_in_pq(new_pq):
    """Return value of highest priority item in priority queue."""
    new_pq.insert(1, 100)
    new_pq.insert(2, 99.9)
    new_pq.insert(3, 99.8)
    assert new_pq.peek() == 1

File: js/test/test-priority-q.js
'use strict';

let Pq = require('../priority-q');
let chai = require('chai');
let expect = chai.expect;

describe('Functions that test Priority class methods.', function() {

  it('Test newly created PQ has default values.', function() {
    let pq = new Pq();
    expect(pq.head).to.be.null;
    expect(pq.tail).to.be.null;
    expect(pq._counter).to.equal(0);
  });

  it('Test insert method adds value to Pq.', function() {
    let pq = new Pq();
    pq.insert(1);
    expect(pq.head.data).to.equal(1);
  });

  it('Test insert method adds value to back of Pq.', function() {
    let pq = new Pq();
    pq.insert(1);
    pq.insert(2);
    pq.insert(3);
    expect(pq.head.data).to.equal(1);
  });

  it('Test insert method adds value and priority to back of Pq.', function() {
    let pq = new Pq();
    pq.insert(1, 100);
    expect(pq.head.priority).to.equal(100);
  });

  it('Test insert method increases size of Pq.', function() {
    let pq = new Pq();
    pq.insert(1);
    expect(pq._counter).to.equal(1);
  });

  it('Test insert method throws error if priority lower zero.', function() {
    let pq = new Pq();
    expect(pq.insert).to.throw(Error);
  });

  it('Test pop method removes value from Pq.', function() {
    let pq = new Pq();
    pq.insert(1);
    expect(pq.pop()).to.equal(1);
  });

  it('Test pop method removes head of Pq.', function() {
    let pq = new Pq();
    pq.insert(1);
    pq.insert(2);
    pq.insert(3);
    expect(pq.pop()).to.equal(1);
  });

  it('Test pop method reduces size of Pq.', function() {
    let pq = new Pq();
    pq.insert(1);
    pq.insert(2);
    pq.insert(3);
    pq.pop();
    expect(pq._counter).to.equal(2);
  });

  it('Test pop method removes highest priority value in Pq.', function() {
    let pq = new Pq();
    pq.insert(1, 20);
    pq.insert(2, 40);
    pq.insert(3, 600);
    expect(pq.pop()).to.equal(3);
  });

  it('Test pop method throws error if Pq is empty.', function() {
    let pq = new Pq();
    expect(pq.pop).to.throw(Error);
  });

  it('Test peek method returns val of head if all values have same priority.', function() {
    let pq = new Pq();
    pq.insert(1);
    pq.insert(2);
    pq.insert(3);
    expect(pq.peek()).to.equal(1);
  });

  it('Test peek method returns highest priority value item in Pq.', function() {
    let pq = new Pq();
    pq.insert(1, 5);
    pq.insert(2, 10);
    pq.insert(3, 6000);
    expect(pq.peek()).to.equal(3);
  });

  it('Test peek method throws error if Pq is empty.', function() {
    let pq = new Pq();
    expect(pq.peek).to.throw(Error);
  });

});
